README.md
# vkeys

## problem
some browsers have got an e.key attribute that is submitted with `keydown` and `keyup` events.
chrome (and other browsers) does not have a e.key attribute.
there are also inconsistencies between different browsers.

## solution

@chrisdickinson initiated a project [vkey](https://github.com/chrisdickinson/vkey) to collect the different definitions and `merge` them into a unified definition.
for code that can not handle `spaces`, or wants less clutter, this transformation might be useful.

this module transforms the [vkey](https://github.com/chrisdickinson/vkey) definition, and __removes__ the following __clutter__:
 - `spaces`
 - `<` and `>`
 - `-`
 - runs `toLowerCase()`

the output is a static key mapping file (without dynamic edge cases): [vkeys.js](vkeys.js)

# usage

## install
```bash
npm install vkeys
```

## use
```js
var vkeys = require('./vkeys');

window.addEventListener('keydown', function(e) {
  console.log(vkeys[e.keyCode])
});
```

# api

#### vkeys

Object with {`code` : `key`} pairs.
```js
{
  0: 'unk',
  ...
  8: 'backspace',
  9: 'tab',
  12: 'clear',
  13: 'enter',
  ...
  254: 'clear'
}
```
see: [vkeys.js](https://github.com/intesso/vkeys/blob/master/vkeys.js)

#### vkeys.getKey(code)

return the `key` for the given `code`.
```js
var key = vkeys.getKey(60);
assert.equal(key, '<');
```
#### vkeys.findCode(key)

returns the first `code` that matches the `key`.
```js
var code = vkeys.findCode('space');
assert.equal(code, 32);
```

#### vkeys.findAllCodes(key)

returns an Array of `code`'s that match the `key`.
```js
var code = vkeys.findAllCodes('meta');
// code equals [91, 92, 223, 224]
```


# build

 1. get the latest vkey module: `npm install --save-dev vkey`
 2. run the build: `npm run build`

# test

```bash
npm install -g browserify testling
npm run test
```

# license

MIT

History.md

1.0.0 / 2015-02-20
==================

 * Initial commit


1.2.0 / 2015-02-20
==================

 * adds utility functions


test/test.findStuff.js
var test = require('tape');
var vkeys = require('../vkeys');


test('should find code 32', function (t) {
  t.plan(1);
  var code = vkeys.findCode('space');
  t.equal(code, 32, 'key space should have code 32');
});

test('should find codes [91,92,223,224]', function (t) {
  t.plan(2);
  var codes = vkeys.findAllCodes('meta');
  console.log('codes', codes);
  t.equal(codes.length, 4);
  var arr = [91, 92, 223, 224];
  t.true(arrayEqual(codes, arr));
});

test('should get key "<"', function (t) {
  t.plan(1);
  var key = vkeys.getKey(60);
  t.equal(key, '<', 'key space should have key "<"');
});


function arrayEqual(actual, expected) {
  var equal = true;
  actual.forEach(function (k) {
    if (actual[k] != expected[k]) equal = false;
  });
  return equal;
}


src/test-keycode.js
// event comparison (outdated): http://unixpapa.com/js/key.html
var events = ['keydown', 'keyup', 'keypress', 'textInput'];
events.forEach(function (event) {
  window.addEventListener(event, function (e) {
    e.preventDefault();
    e.stopPropagation();
    console.log('Event: ', event, ': keyCode', e.keyCode, ', which', e.which, ', charCode', e.charCode);
    console.log('EventObject:', e);
    return false;
  });
});


// event comparison (outdated): http://unixpapa.com/js/key.html
var events = ['keydown'];
events.forEach(function (event) {
  window.addEventListener(event, function (e) {
    e.preventDefault();
    e.stopPropagation();
    console.log('Event: ', event, ': keyCode', e.keyCode, ', which', e.which, ', charCode', e.charCode, ' ,object', e, JSON.stringify(e));
    return false;
  });
});

// firefox has got e.key which is the name of the key. so creating a logger would be helpful.
// e.keyCode is available in firefox and chrome


// so let's create a logger:
var vkey = window.vkey = {};
var events = ['keydown'];
events.forEach(function (event) {
  window.addEventListener(event, function (e) {
    e.preventDefault();
    e.stopPropagation();
    if (!e.key) return console.log('please try with firefox, event.key is not available in your browser');
    vkey[e.keyCode] = e.key;
    console.log('added: ', 'keyCode', e.keyCode, ', key', e.key);
    return false;
  });
});

// let's move on and create a project to collect the keys from different keyboards with different OS around the world.
// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent.key#Key_values
// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent.key#Browser_compatibility


var vkey = window.vkey = {};
var events = ['keydown', 'appcommand'];
events.forEach(function (event) {
  window.addEventListener(event, function (e) {
    e.preventDefault();
    e.stopPropagation();
    if (!e.key) return console.log('please try with firefox, event.key is not available in your browser');
    if (!vkey[e.keyCode]) {
      console.log('yay, you found a new keyCode!');
      vkey[e.keyCode] = {};
    }

    if (!vkey[e.keyCode][e.key]) {
      console.log('ueee, you found a new key for a keyCode!');
      vkey[e.keyCode][e.key] = getContext(e);
      console.log('added: ', 'keyCode', e.keyCode, ', key', e.key);
    }
    return false;
  });
});

function getContext(e) {
  var obj = {};
  obj.keys = getKeys(e);
  obj.userInfo = getUserInfo();
  obj.date = new Date;
  return obj;
}

function getKeys(e) {
  return copy(e, ['ctrlKey', 'altKey', 'metaKey', 'shiftKey']);
}

function getUserInfo() {
  return copy(navigator, ['platform', 'languages', 'userAgent']);
}

function copy(obj, attributes) {
  var copy = {};
  attributes.forEach(function (attribute) {
    copy[attribute] = obj[attribute];
  });
  return copy;
}


// next attempt, only store keys when no other key is pressed ['ctrlKey', 'altKey', 'metaKey', 'shiftKey']
var vkey = window.vkey = {};
var events = ['keydown', 'appcommand'];
events.forEach(function (event) {
  window.addEventListener(event, function (e) {
    e.preventDefault();
    e.stopPropagation();
    if (!e.key) return console.log('please try with firefox, event.key is not available in your browser');
    var other = otherKeyPressed(e);
    if (other) return console.log('only press one key at a time', other, e);

    if (!vkey[e.keyCode]) {
      vkey[e.keyCode] = e.key;
      console.log('added: ', 'keyCode', e.keyCode, ', key', e.key);
    }
    return false;
  });
});

function otherKeyPressed(e) {
  return hasOne(e, ['ctrlKey', 'altKey', 'metaKey', 'shiftKey']);
}

function hasOne(obj, attributes) {
  return attributes.some(function (attribute) {
    return obj[attribute];
  });
}


vkeys.js
var vkeys = exports = module.exports = {
  0: 'unk',
  1: 'mouse1',
  2: 'mouse2',
  3: 'break',
  4: 'mouse3',
  5: 'mouse4',
  6: 'mouse5',
  8: 'backspace',
  9: 'tab',
  12: 'clear',
  13: 'enter',
  16: 'shift',
  17: 'ctrl',
  18: 'alt',
  19: 'pause',
  20: 'capslock',
  21: 'imehangul',
  23: 'imejunja',
  24: 'imefinal',
  25: 'imekanji',
  27: 'escape',
  28: 'imeconvert',
  29: 'imenonconvert',
  30: 'imeaccept',
  31: 'imemodechange',
  32: 'space',
  33: 'pageup',
  34: 'pagedown',
  35: 'end',
  36: 'home',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down',
  41: 'select',
  42: 'print',
  43: 'execute',
  44: 'snapshot',
  45: 'insert',
  46: 'delete',
  47: 'help',
  48: '0',
  49: '1',
  50: '2',
  51: '3',
  52: '4',
  53: '5',
  54: '6',
  55: '7',
  56: '8',
  57: '9',
  58: ':',
  59: ';',
  60: '<',
  61: '=',
  62: '>',
  63: '?',
  64: '@',
  65: 'a',
  66: 'b',
  67: 'c',
  68: 'd',
  69: 'e',
  70: 'f',
  71: 'g',
  72: 'h',
  73: 'i',
  74: 'j',
  75: 'k',
  76: 'l',
  77: 'm',
  78: 'n',
  79: 'o',
  80: 'p',
  81: 'q',
  82: 'r',
  83: 's',
  84: 't',
  85: 'u',
  86: 'v',
  87: 'w',
  88: 'x',
  89: 'y',
  90: 'z',
  91: 'meta',
  92: 'meta',
  93: 'menu',
  95: 'sleep',
  96: 'num0',
  97: 'num1',
  98: 'num2',
  99: 'num3',
  100: 'num4',
  101: 'num5',
  102: 'num6',
  103: 'num7',
  104: 'num8',
  105: 'num9',
  106: 'num*',
  107: 'num+',
  108: 'numenter',
  109: 'num-',
  110: 'num.',
  111: 'num/',
  112: 'f1',
  113: 'f2',
  114: 'f3',
  115: 'f4',
  116: 'f5',
  117: 'f6',
  118: 'f7',
  119: 'f8',
  120: 'f9',
  121: 'f10',
  122: 'f11',
  123: 'f12',
  124: 'f13',
  125: 'f14',
  126: 'f15',
  127: 'f16',
  128: 'f17',
  129: 'f18',
  130: 'f19',
  131: 'f20',
  132: 'f21',
  133: 'f22',
  134: 'f23',
  135: 'f24',
  144: 'numlock',
  145: 'scrolllock',
  160: 'shiftleft',
  161: 'shiftright',
  162: 'ctrlleft',
  163: 'ctrlright',
  164: 'altleft',
  165: 'altright',
  166: 'browserback',
  167: 'browserforward',
  168: 'browserrefresh',
  169: 'browserstop',
  170: 'browsersearch',
  171: 'browserfavorites',
  172: 'browserhome',
  173: 'volumemute',
  174: 'volumedown',
  175: 'volumeup',
  176: 'nexttrack',
  177: 'prevtrack',
  178: 'stop',
  179: 'playpause',
  180: 'launchmail',
  181: 'launchmediaselect',
  182: 'launchapp1',
  183: 'launchapp2',
  186: ';',
  187: '=',
  188: ',',
  189: '-',
  190: '.',
  191: '/',
  192: '`',
  219: '[',
  220: '\\',
  221: ']',
  222: '\'',
  223: 'meta',
  224: 'meta',
  226: 'altgr',
  229: 'imeprocess',
  231: 'unicode',
  246: 'attention',
  247: 'crsel',
  248: 'exsel',
  249: 'eraseeof',
  250: 'play',
  251: 'zoom',
  252: 'noname',
  253: 'pa1',
  254: 'clear'
};

exports.findCode = function findCode(key) {
  for (var k in vkeys) {
    if(vkeys.hasOwnProperty(k)) {
      if (key == vkeys[k]) return parseInt(k);
    }
  }
  return null;
};

exports.findAllCodes = function findAllCodes(key) {
  var codes = Object.keys(vkeys).filter(function (k) {
    return (key == vkeys[k]);
  });
  return codes.map(function(code) {
    return parseInt(code);
  })
};

exports.getKey = function getKey(code) {
  return vkeys[code];
};
